### PR TITLE
[IMP] web: various Popover improvements

### PR DESCRIPTION
**Pull request by seb-odoo**
Close can now be requested by children (slot content) with the event
`o-popover-close`. This feature is necessary if we want a "close" button inside
the popover itself, which is the case for mail.activity "Mark Done".

Clicking on the popover target now acts as a toggle, which is also a feature
that was expected from the previous popover implementation.

Add a specific class (`o_is_open`) when the popover is displayed to be able to
style its target depending on whether it is displayed or not.

Ensure computed position is whole number of pixels to prevent flicker issue on
Firefox.

Remove forced width to let the component grow with its content.

Add appropriate z-index to prevent overlap issue with the rest of our interface.

### CODE REVIEW

File: addons/web/static/src/js/core/popover.js [hunk 1]
```diff
@@ -26,37 +26,75 @@ odoo.define('web.Popover', function () {
             this.state = useState({
                 displayed: false,
             });
+
+            this._onDocumentClick = this._onDocumentClick.bind(this);
+            this._onDocumentScroll = this._onDocumentScroll.bind(this);
+            this._onWindowResize = this._onWindowResize.bind(this);
+
+            this._onDocumentScroll = _.throttle(this._onDocumentScroll, 50);
+            this._onWindowResize = _.debounce(this._onWindowResize, 250);
```
Comment (alexkuhn) [inline addons/web/static/src/js/core/popover.js:35]: prefer naming `onClickDocument`, `onScrollDocument`, etc.

File: addons/web/static/src/js/core/popover.js [hunk 2]
```diff
@@ -26,37 +26,75 @@ odoo.define('web.Popover', function () {
...
             if (!this.state.displayed) {
                 return;
             }
+
             // get the target from the dom
             // we don't want to do this early, since the target might not
-            // be in the dom when the component is instanciated
-            if (!this.constructor._isInViewport(this.el)) {
+            // be in the dom when the component is instantiated
+            if (!this.env.services.isElementInViewport(this.el)) {
```
Comment (alexkuhn) [inline addons/web/static/src/js/core/popover.js:87]: weird to define that as a service, but I suppose this is to easily mock it in tests and remove the hard-code styling of before.

File: addons/web/static/src/js/core/popover.js [hunk 3]
```diff
@@ -90,34 +128,69 @@ odoo.define('web.Popover', function () {
...
+        //----------------------------------------------------------------------
+        // Handlers
+        //----------------------------------------------------------------------
 
         /**
+         * Toggles the popover depending on its current state.
+         *
+         * @param {MouseEvent} ev
+         */
+        _onClick(ev) {
```
Comment (alexkuhn) [inline addons/web/static/src/js/core/popover.js:140]: missing `@private` tag (some other handler functions too)

File: addons/web/static/src/js/core/popover.js (hunk 1 above)
Comment (seb-odoo) [reply addons/web/static/src/js/core/popover.js:35]: Ok

File: addons/web/static/src/js/core/popover.js (hunk 3 above)
Comment (seb-odoo) [reply addons/web/static/src/js/core/popover.js:140]: Done, those were method that were public before, forgot to add it.

File: addons/web/static/src/js/core/popover.js (hunk 2 above)
Comment (seb-odoo) [reply addons/web/static/src/js/core/popover.js:87]: This is how the `web` env is currently set up, "helpers" are basically all in `services`.

As for the goal:
- it shouldn't be in the component itself as it is a fairly generic method
- it could be a helper, but it is harder to test
- -> env is the best place for it

One of the tests is still relying on the hard-coded positioning (the one actually testing the position of the popover in regard to the viewport) but all the other tests indeed don't care about this at all, as long as a popover is shown.

**Comment by seb-odoo** [comment]
@dbo-odoo as discussed, improved Popover for my needs
(will probably need open on hover later but this can be done later)

@aab-odoo will you review this please?

File: addons/web/static/src/js/core/popover.js [hunk 4]
```diff
@@ -26,37 +26,77 @@ odoo.define('web.Popover', function () {
...
             if (!this.state.displayed) {
                 return;
             }
+
             // get the target from the dom
             // we don't want to do this early, since the target might not
-            // be in the dom when the component is instanciated
-            if (!this.constructor._isInViewport(this.el)) {
+            // be in the dom when the component is instantiated
+            if (!this.env.services.isElementInViewport(this.el)) {
```
Comment (aab-odoo) [inline addons/web/static/src/js/core/popover.js:89]: Not sure we want this in the env. Do you think of other usecases? Because for now, we can easily mock a static method on Popover in the tests to achieve what you want.

cc @ged-odoo What do you think?

File: addons/web/static/src/js/core/popover.js [hunk 5]
```diff
@@ -26,37 +26,77 @@ odoo.define('web.Popover', function () {
...
+             * Use capture for the following events to ensure no other part of
+             * the code can stop its propagation from reaching here.
+             */
+            document.addEventListener('click', this._onClickDocument, {
+                capture: true,
+            });
+            document.addEventListener('scroll', this._onScrollDocument, {
+                capture: true,
+            });
+            window.addEventListener('resize', this._onResizeWindow);
```
[the remark this answers is not in the document]
Comment (aab-odoo) [reply addons/web/static/src/js/core/popover.js:49]: Also, with the previous implementation, we only bound these handlers once, no matter how many popovers there were in the DOM. Why do you need to change that?

File: addons/web/static/src/xml/base.xml [hunk 6]
```diff
@@ -508,7 +508,7 @@
 </t>
 
 <t t-name="Popover" owl="1">
-    <div t-on-click="openPopover" data-popover="1">
+    <div t-att-class="{ 'o_is_open': state.displayed }" t-on-click="_onClick" t-on-o-popover-close="_onPopoverClose">
```
Comment (aab-odoo) [inline addons/web/static/src/xml/base.xml:511]: prefixed by 'o-'?

File: addons/web/static/src/js/core/popover.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/web/static/src/js/core/popover.js:49]: I think it's not a good idea to couple different components that are otherwise not aware of each other through the use of global variables.
This change allowed to remove the need of syncing a global variable with the state of the currently displayed component, which in turn made the code of the handlers much easier (it was getting even worse if the close/toggle had to be coupled with it).

If we really want to enforce 1 popover open at a time, there should be a root "popover manager" that properly displays the one that should be. And it would probably need to get this information from a store, as "popover targets" can be placed anywhere in the code, typically not as a children of that manager. While this probably should be done for dialogs, I don't think that's needed for popover.

I didn't know about `useExternalListener` but using it will allow to even further simplify the code, so I will try it out.

File: addons/web/static/src/js/core/popover.js (hunk 4 above)
Comment (seb-odoo) [reply addons/web/static/src/js/core/popover.js:89]: I believe mocking/patching has no place in the world of OWL/env.

If the same method is used both in the window instantiating the tests and in the tests themselves, it is not possible to patch one without affecting the other.

File: addons/web/static/src/xml/base.xml (hunk 6 above)
Comment (seb-odoo) [reply addons/web/static/src/xml/base.xml:511]: Those are DOM events that can potentially conflict with any other lib, similar to what we do for CSS classes.

Livechat will need popover, and it can be included on any website.

**Comment by seb-odoo** [comment]
Updated with `useExternalListener`.

File: addons/web/static/src/js/core/popover.js (hunk 4 above)
Comment (aab-odoo) [reply addons/web/static/src/js/core/popover.js:79]: > I believe mocking/patching has no place in the world of OWL/env.
 
I agree with you on that. However, I perceive tests a little bit differently. I believe it's acceptable to do that kind of stuff in tests, occasionally.

> If the same method is used both in the window instantiating the tests and in the tests themselves, it is not possible to patch one without affecting the other.

Not sure to understand what you mean, but if you mock a static method on Popover, it's very unlikely to be used by the test engine.

I'm not saying you shouldn't do it. As I said, I don't really know. But I'm not comfortable with the idea of having such functions in the env. I think that what bothers me is that it is obvious that i will be overridden to hack the result, not because it's something you want to configure in your app, or in a subtree.

Anyway, I'd like to have @ged-odoo opinion on this.

File: addons/web/static/src/xml/base.xml (hunk 6 above)
Comment (aab-odoo) [reply addons/web/static/src/xml/base.xml:511]: good point

File: addons/web/static/src/js/core/popover.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (aab-odoo) [reply addons/web/static/src/js/core/popover.js]: We may have been overzealous, but with list or kanban views, we can quickly reach a larger number of popovers into the DOM, thus leading to a larger number of (global) event handlers. It's easy to imagine a situation where you have 80 records in a list, with a widget having a popover. In this case, every click anywhere in the app, will trigger the execution of 80 handlers. Same for scroll (and less frequently, resize). So I will maybe not lead to performance issues, but it's not elegant at all. And you can still change the limit, so even performance-wise it could get bad.

Also, I agree with you about the PopoverManager. I think it would be excessive.

Actually, we already had this reflection before merging the initial PR, and we considered those 3 solutions. We went for the middle one, which offered the best compromise between performance, and simplicity (having the whole Popover stuff handled by a single Component).

[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/web/static/src/js/core/popover.js]: Ok, I didn't realize through the comments that it was such an important optimization, and it wasn't covered by tests either.

But now I that understand better, I think I know how to do it to keep that aspect while still avoiding global variables.

File: addons/web/static/src/js/core/popover.js (hunk 4 above)
Comment (seb-odoo) [reply addons/web/static/src/js/core/popover.js:79]: > I think that what bothers me is that it is obvious that i will be overridden to hack the result, not because it's something you want to configure in your app, or in a subtree.

This is true, but if this ends up being the decisive criteria, note that it is also true for almost all the other things we put in the env services so far (the only exception seems to be `httpRequest` where we actually want a different behavior on the public side).

But I think tests should not be considered differently than configuring app. In regard to the components themselves, they should not be aware whether they are tested or not, they are just instantiated in another env, literally.

Using the `env` approach also make very explicit what is expected to change depending on the situation. With mock or patch, you never really know what is actually hacked or not, and it can make debugging (tests) issues harder (forgot to unpatch in afterEach which is impacting following tests, etc. None of this is necessary with env).

**Comment by seb-odoo** [comment]
Updated PR with smart (un)binding of global events (only if popover currently open).

File: addons/web/static/src/js/core/popover.js (hunk 4 above)
Comment (ged-odoo) [reply addons/web/static/src/js/core/popover.js:105]: @seb-odoo @aab-odoo 
I think that we should use an API that is designed exactly to solve this very issue: https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API.
This is also likely to be way more efficient in all cases.

Also, this is not really a service, but an access to a browser API. As such, I think that this should be located in `env.browser`.

- `env.browser` is where we add everything that is simply an access to native browser API, such as localStorage, Math.random, fetch, ...
- `env.services` is the location of higher level of abstraction pieces that needs to be alive for the duration of the application.  So, mail service for example. Services need to use `env.browser` as well if needed, so it can easily be mocked.

Comment (seb-odoo) [reply addons/web/static/src/js/core/popover.js:105]: This is interesting, but I'm not sure if it fits the need?

The current code does not really need an observer, as it doesn't need to be notified of element potentially getting out of view at some point in the future.

It is only checking if the currently computed position fits on the viewport at the particular time of the compute, to decide if another position has to be computed (if outside) or if it's fine to keep it as a valid position.
There's also the problem that if it's completely outside, no callback will ever be run apparently, except we don't want to wait indefinitely for a callback, we just want to know it's outside.

So changing that seems out of scope of the current PR, and maybe not needed at all.

With that in mind, what is left to do here? Make new `browser` env key and move the current method there instead of in `services`?

**Comment by seb-odoo** [comment]
Updated with `isElementInViewport` in `browser` instead of `services`.

**Comment by ged-odoo** [comment]
> Updated with `isElementInViewport` in `browser` instead of `services`.

oh no!  `browser` is meant to contain only native browser API, not a bunch of helper functions.  In your case, if the intersection observer does not help you, then please do a simple function in your component, and patch it in the test. This is way more simple.

**Comment by seb-odoo** [comment]
Removed the change in `env` and positioned the fixture manually in the viewport in the new tests, similar to what was done in existing tests.

**Comment by seb-odoo** [comment]
Any news about this PR?

I have just added 2 small CSS fixes to it:

- Remove forced width to let the component grow with its content.
- Add appropriate z-index to prevent overlap issue with the rest of our interface.

**Comment by seb-odoo** [comment]
Also meant to include the following change, now done too:

- Ensure computed position is whole number of pixels to prevent flicker issue on
Firefox.

**Comment by seb-odoo** [comment]
Also a note, in regard to the forced width change: if the content of the popover is expected to change in size (which would include height too btw) there should probably be a mechanism in place (trigger?) to ensure the popover is properly computing itself again if needed. However, this might cause more pain than it is worth (flicker, stuff moving around constantly, ...). I'm not sure if there is an ideal solution here, so I would probably consider it out of scope, but I feel we should be aware of it.

File: addons/web/static/src/scss/popover.scss [hunk 7]
```diff
@@ -52,12 +52,12 @@
     $arrow-pad: 8px;
     $arrow-size: 8px;
     $animation-time: 0.2s;
-    width: 280px;
     max-width: 100vw;
 
     position: fixed;
     top: 0; // will be updated in js
     left: 0; // but let's set it in the viewport in case of test/debug
+    z-index: 1060;
```
Comment (aab-odoo) [inline addons/web/static/src/scss/popover.scss:60]: we should maybe have a variable for this

File: addons/web/static/src/js/core/popover.js [hunk 8]
```diff
@@ -26,37 +26,84 @@ odoo.define('web.Popover', function () {
...
+             * Those events are only necessary if the popover is currently open,
+             * so we decided for performance reasons to avoid binding them while
+             * it is closed. This allows to have many popover instantiated while
+             * keeping the count of global handlers low.
+             */
+            this._hasGlobalEventListeners = false;
+        }
+
+        mounted() {
+            this._compute();
```
Comment (aab-odoo) [inline addons/web/static/src/js/core/popover.js:47]: No call to super when you directly inherit from Component? I thought we agreed to always call super.

File: addons/web/static/src/js/core/popover.js [hunk 9]
```diff
@@ -70,7 +117,14 @@ odoo.define('web.Popover', function () {
                 .find((pos) => {
                     this.popoverRef.el.style.top = `${pos.top}px`;
                     this.popoverRef.el.style.left = `${pos.left}px`;
-                    return this.constructor._isInViewport(this.popoverRef.el);
+                    const rect = this.popoverRef.el.getBoundingClientRect();
```
Comment (aab-odoo) [inline addons/web/static/src/js/core/popover.js:120]: I though you needed to be able to override this in tests. A static function _isInViewPort that can be overridden to always return true in tests was the easiest way to achieve what you wanted, imho. But I may have misunderstood the needs.

**Comment by dbo-odoo** [comment]
@seb-odoo In this case, I think the bootstrap way of thinking was better: having a fixed width makes more sense, as it gets ugly in a lot of case otherwise.
screenshot-2611822-9517-6ba0db runbot43 odoo com-2020 05 13-16_03_50

Is there a motivation behind this particular change, besides "why not"? Genre tu construis la porte du 2ème étage au cas ou on veut y accéder ou bien ça te pose un vrai souci?

**Comment by dbo-odoo** [comment]
I'd prefer having a fixed width by default and being able to disable it when i want rather than the opposite (esp. since `Popover` is in the component registry, which means it should behave sensibly out of the box).

File: addons/web/static/src/scss/popover.scss [hunk 10]
```diff
@@ -52,12 +52,12 @@
     $arrow-pad: 8px;
     $arrow-size: 8px;
     $animation-time: 0.2s;
-    width: 280px;
```
Comment (aab-odoo) [inline addons/web/static/src/scss/popover.scss:55]: Not sure about this (but the fixed width seems to be a bad idea as well). Here, what prevents from having a super long single line popover?

**Comment by seb-odoo** [comment]
The main problem I had with the width is that I wanted it to be _smaller_, not larger. It was just way too ugly with all that empty space in my case (old popover didn't have that issue, it was smallest possible by default).

I don't really mind if the width if fixed by default or not, but it should be possible to say "I don't want it" in a way that does not imply having to mess around with CSS overrides.

So basically I'm ok if the width becomes a prop that has 280 as default value and that accepts undefined/0 as no specific width, if that is what we prefer.

But with that in mind maybe it should be some kind of max-width instead, because it seems in your example that you don't want it to become too big? And me I just want to have it as small as possible.

**Comment by dbo-odoo** [comment]
i think a prop is perhaps the way to go, though usually other widgets/components have a `'small', 'medium', 'large'` selection instead of a pixel count, which I think is preferrable.

File: addons/web/static/src/js/core/popover.js (hunk 9 above)
Comment (seb-odoo) [reply addons/web/static/src/js/core/popover.js:120]: I don't need to override it in the tests at the end, because when it matters, the test is positioning the element in the viewport manually instead. Basically I just reverted to the old code of the tests since you guys weren't too excited with the env change, I thought it would be an improvement but I just discarded it.

The only difference is that previously the tests that didn't care about the positioning also needed the element to be in viewport to show the popover, and now they don't care. Instead of complaining that the target is not in viewport, the component is just picking the default position instead. That change also fixed a bug where during some resize situations, the popover would briefly go outside of viewport and disappear even though it should eventually show again, which is does now.

File: addons/web/static/src/js/core/popover.js (hunk 8 above)
Comment (seb-odoo) [reply addons/web/static/src/js/core/popover.js:47]: We never use super in our refactoring so that must have been the original guideline. I have no idea now, please let us know if this changed. But as far as I can see, all components that have been merged so far in master don't call super either.

File: addons/web/static/src/scss/popover.scss (hunk 7 above)
Comment (seb-odoo) [reply addons/web/static/src/scss/popover.scss:60]: Agree, there was with bootstrap: `$zindex-popover: 1060 !default;`

We can either use it again or copy/paste it somewhere and prefix it with `-o`.

**Comment by seb-odoo** [comment]
Predefined sizes make sense for dialogs since they are on top of everything else. And they would actually be ugly if too small. Though at the opposite we do have cases where even large dialog is not enough (typically when media are involved to give one example).

But for popover as opposed to dialog it is supposed to be a rather small contextual interface element where the smaller it is, the better.
Bootstrap popover didn't have any minimal size, I think it was sound and I would like to keep it this way.

Do we need to brainstorm this on a channel, or can we come to an agreement as to how to do it here? Using text takes away more time than I feel should be necessary for this.

**Comment by seb-odoo** [comment]
After discussing with @dbo-odoo we share the same view on what should (not) be done: either no width restriction, or maybe a default max-width for convenience. A fixed width like before this PR is definitely not what we want to keep.

@aab-odoo @kebeclibre @ged-odoo would any (or all) of you be available tomorrow or this Friday to iron this out?

**Comment by aab-odoo** [comment]
@seb-odoo @dbo-odoo let's close this one tomorrow (Thursday)

File: addons/web/static/src/js/core/popover.js (hunk 8 above)
Comment (ged-odoo) [reply addons/web/static/src/js/core/popover.js:47]: @aab-odoo  I thought we agreed that it was not necessary on component methods?

**Comment by ged-odoo** [comment]
> I'd prefer having a fixed width by default and being able to disable it when i want rather than the opposite (esp. since `Popover` is in the component registry, which means it should behave sensibly out of the box).

I really think that it should have fixed width. Simple is always better.  Don't overengineer stuff

**Comment by ged-odoo** [comment]
> i think a prop is perhaps the way to go, though usually other widgets/components have a `'small', 'medium', 'large'` selection instead of a pixel count, which I think is preferrable.

strongly agree with this

File: addons/web/static/src/js/core/popover.js [hunk 11]
```diff
@@ -70,7 +117,14 @@ odoo.define('web.Popover', function () {
                 .find((pos) => {
                     this.popoverRef.el.style.top = `${pos.top}px`;
                     this.popoverRef.el.style.left = `${pos.left}px`;
```
Comment (ged-odoo) [inline addons/web/static/src/js/core/popover.js:119]: Now that I read the code, I understand why intersectionobserver is not what you want. But I don't like the way you do it: you randomly try to position the popover and measure it in the dom (which causes a css redraw), then reposition it until it works.

To be honest, I think that you are solving the issue in the wrong way. This is a common issue, and the solution has always been to do some computations in JS (usually based on fixed width/height/current top/left position ), and then apply the style just once.

**Comment by seb-odoo** [comment]
After reporting the various issue to PO team and in particular to NGR he decided that for this version we try to stay as close as possible to the behavior of popover in bootstrap.

It is acceptable to deviate on small details (for example not supporting hover), however the size of the popover is considered a crucial element that should be kept minimal: the goal of popover being to display small contextual information while taking the least amount of space (otherwise a dialog would be more appropriate).

The goal of this PR is to fix small usability issues, while keeping in mind that we might want to refactor the implementation later (in particular in regard to pre-computing the position, which would be a better approach if we find a way to keep the functional aspect intact).

@aab-odoo could you please make a final check so that I don't merge my own PR?

cc @dbo-odoo if you want to also make a final check, though I don't think I changed anything since last time you tested.

**Comment by aab-odoo** [comment]
robodoo r+
